enclave-protocol/src/codec.rs

```diff
@@ -0,0 +1,55 @@
+use std::{
+    convert::TryInto,
+    io::{self, Read, Write},
+    vec,
+    vec::Vec,
+};
+
+use parity_scale_codec::{Decode, Encode};
+
+/// Trait for writing length encoded valus to a stream
+pub trait StreamWrite {
+    fn write_to<W: Write>(&self, writer: W) -> io::Result<usize>;
+}
+
+/// Trait for reading values written using `StreamWrite`
+pub trait StreamRead: Decode {
+    fn read_from<R: Read>(reader: R) -> io::Result<Self>;
+}
+
+impl<T: Encode> StreamWrite for T {
+    fn write_to<W: Write>(&self, mut writer: W) -> io::Result<usize> {
+        let mut bytes = self.encode();
+        let mut to_send = Vec::with_capacity(
+            bytes
+                .len()
+                .checked_add(8)
+                .expect("Bytes to send cannot fit in a `Vec`"),
+        );
+
+        let size: u64 = bytes
+            .len()
+            .try_into()
+            .expect("Bytes to send cannot fit into u64");
+
+        to_send.extend(&size.to_le_bytes());
+        to_send.append(&mut bytes);
+
+        writer.write(&to_send)
+    }
+}
+
+impl<T: Decode> StreamRead for T {
+    fn read_from<R: Read>(mut reader: R) -> io::Result<Self> {
+        let mut size = [0u8; 8];
+        reader.read_exact(&mut size)?;
```

Comment:
would there be any use case for transferring 18.446744073709553 _Exabytes_ ? or why u64 lengths without any sanity checks?

Reply:
Changed from `u64` to `u32`.